src/storage/concurrency_manager/mod.rs

```diff
@@ -0,0 +1,171 @@
+    },
+};
+use txn_types::{Key, TimeStamp};
+
+// TODO: Currently we are using a Mutex<BTreeMap> to implement the handle table.
+// In the future we should replace it with a concurrent ordered map.
+// Pay attention that the async functions of ConcurrencyManager should not hold
+// the mutex.
+pub type DefaultConcurrencyManager = ConcurrencyManager<Mutex<BTreeMap<Key, Arc<KeyHandle>>>>;
+
+pub struct ConcurrencyManager<M: OrderedMap> {
+    max_read_ts: Arc<AtomicU64>,
+    handle_table: HandleTable<M>,
+}
+
+impl<M: OrderedMap> Clone for ConcurrencyManager<M> {
+    fn clone(&self) -> Self {
+        ConcurrencyManager {
+            max_read_ts: self.max_read_ts.clone(),
+            handle_table: self.handle_table.clone(),
+        }
+    }
+}
+
+impl<M: OrderedMap> ConcurrencyManager<M> {
+    pub fn new(latest_ts: TimeStamp) -> Self {
+        ConcurrencyManager {
+            max_read_ts: Arc::new(AtomicU64::new(latest_ts.into_inner())),
+            handle_table: HandleTable::default(),
+        }
+    }
+
+    pub fn max_read_ts(&self) -> TimeStamp {
+        TimeStamp::new(self.max_read_ts.load(Ordering::SeqCst))
+    }
+
+    /// Acquires a mutex of the key and returns an RAII guard. When the guard goes
+    /// out of scope, the mutex will be unlocked.
+    ///
+    /// The guard can be used to store LockInfo in the table. The stored lock
+    /// is visible to `read_key_check` and `read_range_check`.
+    pub async fn lock_key(&self, key: &Key) -> KeyHandleMutexGuard<'_, M> {
+        self.handle_table.lock_key(key).await
+    }
+
+    /// Acquires mutexes of the keys and returns the RAII guards. The order of the
+    /// guards is the same with the given keys.
+    ///
+    /// The guards can be used to store LockInfo in the table. The stored lock
+    /// is visible to `read_key_check` and `read_range_check`.
+    pub async fn lock_keys(
+        &self,
+        keys: impl Iterator<Item = &Key>,
+    ) -> Vec<KeyHandleMutexGuard<'_, M>> {
+        let mut keys_with_index: Vec<_> = keys.enumerate().collect();
+        // To prevent deadlock, we sort the keys and lock them one by one.
+        keys_with_index.sort_by_key(|(_, key)| *key);
+        let mut result: Vec<MaybeUninit<KeyHandleMutexGuard<'_, M>>> = Vec::new();
+        result.resize_with(keys_with_index.len(), || MaybeUninit::uninit());
+        for (index, key) in keys_with_index {
+            result[index] = MaybeUninit::new(self.handle_table.lock_key(key).await);
+        }
+        #[allow(clippy::unsound_collection_transmute)]
+        unsafe {
+            mem::transmute(result)
+        }
+    }
+
+    /// Checks if there is a memory lock of the key which blocks the read.
+    /// The given `check_fn` should return false iff the lock passed in
+    /// blocks the read.
+    ///
+    /// It will also updates the max_read_ts.
+    pub fn read_key_check(
+        &self,
+        key: &Key,
+        ts: TimeStamp,
+        check_fn: impl FnOnce(&LockInfo) -> bool,
```

Comment:
Why do we need to take `check_fn`? Rather than just considering a key locked if the lock exists?